### ISSUE TITLE
Timeouts and deadlocked threads

### ISSUE THREAD

**Issue by not-sungard** [arangodb/arangodb#7411]
## My Environment

* __ArangoDB Version__:        3.3.2
* __Storage Engine__:             RocksDB
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   Windows service
* __Configuration__:               Physical, 32-core (16 physical)
* __Infrastructure__:               own
* __Operating System__:        Windows Server 2012 R2
* __Used Package__:              Windows installer

## Component, Query & Data

__Affected feature__:
Using the HTTP REST interface directly, no driver.
Document collections only.

## Steps to reproduce

Not simple to immediately reproduce
1. Create collections (say ~1000)
2. Populate with data (500 - 200,000 documents each)
3. Delete collections
4. Goto 1 until requests start timing out

__Problem__:
After the server has been running for sometime (initially over a month) we start to see some requests (POST, PUT, GET) timeout.
Investigating we noticed the ArangoDB server had a very large number of threads running (> 25,000).
A wait chain analysis suggested that a great many of these threads were deadlocked.
We restarted the server. After a few days timeouts started to happen again and we noticed that the number of threads had grown to above 1,500. Again these looked deadlocked.

__Expected result__:
No timeouts and no deadlocked threads.

We are upgrading to ArangoDB 3.3.19.
Will report if the problem is reproducible in that version as well.

**Comment by Simran-B** [comment]
Hi @not-sungard, thanks for the report. I have a few questions:

- Are the collections created in parallel? If so, how many? All at once?
- Do you first create all 1k collections and then start to populate them, or do you create collections and insert documents in an interleaving fashion?
- Is the request to delete a collection issued while documents are still being inserted into this very collection, or is the collection possibly still busy (disk i/o, indexes etc.)?
- Are new collections created with previously used names?

**Comment by not-sungard** [comment]
- Created in parallel, but only a few at a time
- Create collection then start to insert
- Deletion is done much later as part of house keeping. Older data that isn't needed any more.
- They shouldn't be, we are using UUIDs for collection names.

Our environment is a smallish grid, with a queue of "jobs" to run. The number that can run at any one time is constrained by the number of cores. Number of jobs at any moment in range 2 - 10. Each job creates and writes to its own collection.

**Comment by mpoeter** [comment]
@not-sungard can you create a minidump of the deadlocked process using one of the methods described here https://github.com/arangodb/arangodb/blob/devel/README_maintainers.md#windows-debugging ?

**Comment by kunalak** [comment]
@not-sungard In addition to request for minidump by mpoeter , can you also share if this problem is still also seen in version 3.3.19 or 3.4 RC4? (and is occuring only on Windows OS?)

**Comment by not-sungard** [comment]
I have now seen the deadlock issue on my developer workstation.

I have attached a screenshot:
![screenshot](https://user-images.githubusercontent.com/34069444/49372452-98bf6a00-f6f2-11e8-96ff-e599c28c6dd1.png)

I also have a process dump of arangod.exe. Zipped this is ~208MB and too large to attach.

My environment:
ArangoDB Version: 3.3.19
Storage Engine: RocksDB
Deployment Mode: Single Server
Deployment Strategy: Windows service
Configuration: Physical, 24-core (12 physical)
Infrastructure: own
Operating System: Windows 10 Enterprise
Used Package: Windows installer

**Comment by Simran-B** [comment]
@not-sungard Did you close the issue on purpose or accidentally?

**Comment by not-sungard** [comment]
Oops! Accidentally!

**Comment by not-sungard** [comment]
Re-opening

**Comment by dothebart** [comment]
Hi, 
could you inspect the dump like this: 
`cdb -z ./a.dmp -y 'SRV*e:\symbol_cache*https://download.arangodb.com/symsrv_arangodb33;SRV*e:\symbol_cache\ms*http://msdl.microsoft'`
and in cdb you may invoke it printing backtraces like this: 

```
kp (for the backtrace of the current thread)
~*kb (for the backtraces of all threads) 
```
in general it would be good if you could share the dump file with us via i.e. dropbox.

**Comment by jsteemann** [comment]
@not-sungard : Looking forward to receiving backtraces for the deadlocked process...
In the meantime, can you confirm that the storage engine in use is actually RockDB (asking because the default storage engine in the 3.3 branch is MMFiles), if the server still shows a very high number of threads, and if there is anything suspicious in the `arangod.log` log file?
Thanks!

**Comment by not-sungard** [comment]
Yes, RocksDB.
I have sent a link to a secure download the dump file via another channel.
There is nothing suspicious in the arangod.log file.

**Comment by jsteemann** [comment]
Ok, thanks. When the issue occurred last, do you know if the number of threads again was very high, or if it has been in the normal range?

**Comment by not-sungard** [comment]
Yes, on previous occasions we saw the timeouts the thread count was also very high.
I suspect that the reason for the timeout is that the communication thread gets deadlocked.

**Comment by ObiWahn** [comment]
[windows-dump.log](https://github.com/arangodb/arangodb/files/2662403/windows-dump.log)

**Comment by jsteemann** [comment]
According to the dump [here](https://github.com/arangodb/arangodb/issues/7411#issuecomment-445742555) it seems that many threads (1955) are hanging in shutdown (`~Thread()`) and cannot make any progress. This also explains why the number of threads is very high.
Root cause still unknown, investigating further.

**Comment by jsteemann** [comment]
We have an idea what is causing this problem, and are working on a fix right now.

**Comment by jsteemann** [comment]
Fixed in 3.3.21

**Comment by not-sungard** [comment]
That's great news. Thanks to all.

**Comment by OmarAyo** [comment]
Hi @not-sungard 

I would like to inform you that we have released ArangoDB Version 3.3.21 that will hopefully fix your issue. 

Please give it a try and let us know.

Best,

**Comment by not-sungard** [comment]
Hi OmarAyo,

Thanks. Could you confirm that this fix is also in 3.4.1?

All the best.

**Comment by jsteemann** [comment]
Hi @not-sungard,
yes, the fix is also included in 3.4.1, introduced by commit 9dbbd1cc9f3423c7e622b36a7b45dc10a6a9732d.

**Comment by not-sungard** [comment]
That's great, we'll get on with our testing.
Thanks.

**Comment by vinaypyati** [comment]
Hi @not-sungard,

Just sending a quick follow up. Did you have a chance to test the fix? Let us know your feedback. 

Thanks!

**Comment by dothebart** [comment]
Hi, 
most probably windows can't keep up with linux in throughput on similar hardware - thus you will have to provide more resources to do the same thing. 
The best ways to move forward here is to create a test agent that resembles your usage scenario and creates some of your typical load on the database like I demonstrated in this blogpost series:
https://www.arangodb.com/2017/10/performance-analysis-pyarango-usage-scenarios/

If you then could compare the resource utilization between the both. 
Tbh, while I know how to acquire these system figures on linux (as I wrote in that blogpost), I don't know how to do this on windows. 

However, since you confirmed the original issue seems to be fixed, I think we should close this issue, and continue with such a discussion on another ticket?

**Comment by not-sungard** [comment]
Yes, I think that's fair, the original issue appears to be fixed.

Just on speculation, have you done any comparative testing on the same hardware of Linux vs Windows?
If you have any data or observation that you're willing to share I would be very interested.

**Comment by mpoeter** [comment]
The best way to acquire such figures on Windows is via the "Performance Monitor" (perfmon.exe). See e.g. https://www.digitalcitizen.life/basics-about-working-performance-monitor

**Comment by dothebart** [comment]
Yes, exactly what @Simran-B simran was posting. I had been evaluating locking performance on different os/architectures. And the results on the windows machine stood _that far out_ I couldn't believe it and popped in a knoppix.org to check back - and yes, it was drastic. 
Now these are only locks. However they are used all over the place if resources are shared, but must be available for exclusive access. So you see, this factor alone grains sand into the gears. 
The days when there were Windows machines in the Top 500 supercomputers are also over: https://fossbytes.com/top500-most-powerful-supercomputer-list-2018-powered-by-linux/ - most probably for a good reason.

**Comment by not-sungard** [comment]
Thanks, those are all quite interesting.

At present I think that the issues we're having are to do with our corporate standard build/deploy for Windows 10/Server 2019. We're trying to get some modifications to that for our ArangoDB server, but it is proving difficult.

Still, it's good that we can fallback to only recommending Linux deployments internally. Also would be much cheaper for cloud environments.

**Comment by dothebart** [comment]
If you have recommendations what to change in the installer to improve the deployability, let us know. While many of the Avocado-Gang have experinces running operations on *nix based systems, on windows its mostly "run active directory" - The nsis installer should have all available to install it headless.

**Comment by mpoeter** [comment]
I have to admit, I don't really buy Windows performing so much worse. Don't get me wrong, I am not doubting your test results. Quite the opposite - I think it would be interesting to investigate in more detail the reasons for the performance difference. For example, since the tests were performed in February 2015 I assume you used Visual Studio 2013. I know from experience (learned the hard way 😉) that one of the early std::atomic implementations in VS was complete crap. The std::mutex implementation has also been revised in later versions.

My point is that there are more factors involved than simply the OS; the compiler and STL implementation also play a crucial role. For example, a MinGW build would be interesting for comparison as well.

A lot has happened since those tests, so it would be interesting to see some performance numbers with more recent compilers. Nevertheless, this remains a purely synthetic benchmark (interesting, but with limited significance).

What would be _really_ interesting is a performance comparison of ArangoDB in action.

**Comment by dothebart** [comment]
well, pick some of the examples in the series: https://www.arangodb.com/2017/10/performance-analysis-pyarango-usage-scenarios/ and find out ;)

**Comment by not-sungard** [comment]
Adding more feedback on testing this fix under Windows.

The performance differences we see between Linux & Windows are much larger than what the above articles might lead you to expect. I don't want to get into that though for the moment as it isn't the most important issue we're experiencing.

We are seeing the ArangoDB server, on Windows, "freeze" and become completely unresponsive for extended periods of time. When this happens new connection attempts fail and any requests underway then fail to respond (timeout). When in this state the ArangoDB Windows service is also unresponsive and attempting to stop the service leaves it stuck in the "stopping" state. Unlike previously there are not large numbers of deadlocked threads.

It probably not related to this bug except that we have come across it while testing the fix. We will likely raise this as a new issue in the coming days. Hopefully with some more details.
